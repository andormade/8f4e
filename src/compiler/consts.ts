export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
