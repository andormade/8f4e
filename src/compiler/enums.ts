export const enum NameSection {
	FUNCTION_NAME = 0x01,
	LOCAL_NAME = 0x02,
}

export const enum ImportDesc {
	MEMORY = 0x02,
}

export const enum ExportDesc {
	FUNC = 0x00,
}

export const enum Type {
	I32 = 0x7f,
	F32 = 0x7d,
}

export const enum Instruction {
	IF = 0x04,
	ELSE = 0x05,
	END = 0x0b,
	CALL = 0x10,
	LOCAL_GET = 0x20,
	I32_LOAD = 0x28,
	F32_LOAD = 0x2a,
	I32_STORE = 0x36,
	I32_CONST = 0x41,
	F32_CONST = 0x43,
	I32_EQ = 0x46,
	I32_GE_S = 0x4e,
	I32_ADD = 0x6a,
	I32_SUB = 0x6b,
	I32_MUL = 0x6c,
	I32_DIV_U = 0x6e,
	F32_STORE = 0x83,
	F32_ADD = 0x92,
	F32_SUB = 0x93,
	F32_MUL = 0x94,
	F32_DIV = 0x95,
}

export const enum Section {
	CUSTOM = 0x00,
	TYPE = 0x01,
	IMPORT = 0x02,
	FUNCTION = 0x03,
	MEMORY = 0x05,
	EXPORT = 0x07,
	CODE = 0x0a,
}
