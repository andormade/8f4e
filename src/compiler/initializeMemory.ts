const generateOutputAddressLookup = function (compiledModules) {
	const lookup = {};
	compiledModules.forEach(({ outputs, moduleId }) => {
		lookup[moduleId] = outputs.map(output => ({ outputId: output.id, memoryAddress: output.address }));
	});
	return lookup;
};

export const setInitialMemory = function (memory: any, initialMemory: any) {
	for (let i = 0; i < initialMemory.length; i++) {
		memory[i] = initialMemory[i];
	}
};

export const initializeMemory = function (compiledModules, connections) {
	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const memoryBuffer = new Int32Array(memoryRef.buffer);

	const initialMemory = compiledModules.map(({ initialMemory }) => initialMemory).flat();

	setInitialMemory(memoryBuffer, initialMemory);

	const outputAddressLookup = generateOutputAddressLookup(compiledModules);

	return { memoryRef, memoryBuffer, outputAddressLookup };
};
