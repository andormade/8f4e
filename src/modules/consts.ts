export const MODULE_WIDTH_S = 18;
export const MODULE_WIDTH_M = MODULE_WIDTH_S * 2;
export const MODULE_WIDTH_L = MODULE_WIDTH_S * 3;
export const MODULE_WIDTH_XXL = MODULE_WIDTH_S * 14;
export const MODULE_HEIGHT_S = 9;
export const MODULE_HEIGHT_M = MODULE_HEIGHT_S * 2;
export const MODULE_HEIGHT_L = MODULE_HEIGHT_S * 3;
