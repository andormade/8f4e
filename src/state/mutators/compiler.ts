import { State } from '../types';

export default function compiler(state: State, events) {
	const worker = new Worker(new URL('../../../packages/worker/src/index.ts', import.meta.url));
	// @ts-ignore shared: true
	const memoryRef = new WebAssembly.Memory({ initial: 1, maximum: 1, shared: true });

	async function recompile() {
		worker.postMessage({ memoryRef, modules: state.modules, connections: state.connections });
		state.compiler.isCompiling = true;
		state.compiler.lastCompilationStart = performance.now();
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'compilationDone':
				state.compiler.memoryBuffer = new Int32Array(memoryRef.buffer);
				state.compiler.outputAddressLookup = data.payload.outputAddressLookup;
				state.compiler.isCompiling = false;
				const end = performance.now() - state.compiler.lastCompilationStart;
				state.compiler.compilationTime = end.toFixed(2);

				state.modules.forEach(module => {
					if (state.moduleTypes[module.type].transformer) {
						state.moduleTypes[module.type].transformer(
							module,
							state.compiler.memoryBuffer,
							state.compiler.outputAddressLookup
						);
					}
				});
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', recompile);
	events.on('deleteConnection', recompile);
	events.on('addModule', recompile);
	events.on('deleteModule', recompile);
	events.on('init', recompile);

	return () => {
		events.off('createConnection', recompile);
		events.off('deleteConnection', recompile);
		events.off('addModule', recompile);
		events.off('deleteModule', recompile);
		events.off('init', recompile);
	};
}
