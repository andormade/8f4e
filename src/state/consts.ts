import { midiNoteToInt16 } from './helpers/midi';
import { Note } from '../midi/enums';

export const moduleTypes = {
	splitter: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'in', x: 5, y: 20, isInput: true },
			{ id: 'out1', x: 85, y: 20 },
			{ id: 'out2', x: 85, y: 35 },
			{ id: 'out3', x: 85, y: 50 },
			{ id: 'out4', x: 85, y: 65 },
		],
		name: 'Splitter',
		switches: [],
		sliders: [],
		steppers: [],
	},
	sequentialSwitch: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'out', x: 85, y: 20 },
			{ id: 'in1', x: 5, y: 20, isInput: true },
			{ id: 'in2', x: 5, y: 35, isInput: true },
			{ id: 'in3', x: 5, y: 50, isInput: true },
			{ id: 'in4', x: 5, y: 65, isInput: true },
			{ id: 'clock', x: 5, y: 80, isInput: true },
		],
		name: 'Sequential switch',
		switches: [],
		sliders: [],
		steppers: [],
	},
	clockGenerator: {
		width: 100,
		height: 100,
		connectors: [{ id: 'out', x: 85, y: 20 }],
		name: 'Clock generator',
		switches: [],
		sliders: [{ id: 'rate', x: 10, y: 20, width: 10, height: 50, minValue: 0, maxValue: 3000, resolution: 10 }],
		defaultValues: {
			rate: 10,
		},
		steppers: [],
	},
	quantizer: {
		width: 200,
		height: 100,
		connectors: [
			{ id: 'in', x: 5, y: 20, isInput: true },
			{ id: 'out', x: 185, y: 20 },
		],
		switches: [
			{ id: 'note1', onValue: midiNoteToInt16(Note.C0), offValue: -2147483648, x: 40, y: 42, width: 10, height: 10 },
			{
				id: 'note2',
				onValue: midiNoteToInt16(Note.C0_SHARP),
				offValue: -2147483648,
				x: 50,
				y: 30,
				width: 10,
				height: 10,
			},
			{ id: 'note3', onValue: midiNoteToInt16(Note.D0), offValue: -2147483648, x: 60, y: 42, width: 10, height: 10 },
			{
				id: 'note4',
				onValue: midiNoteToInt16(Note.D0_SHARP),
				offValue: -2147483648,
				x: 70,
				y: 30,
				width: 10,
				height: 10,
			},
			{ id: 'note5', onValue: midiNoteToInt16(Note.E0), offValue: -2147483648, x: 80, y: 42, width: 10, height: 10 },
			{ id: 'note6', onValue: midiNoteToInt16(Note.F0), offValue: -2147483648, x: 90, y: 42, width: 10, height: 10 },
			{
				id: 'note7',
				onValue: midiNoteToInt16(Note.F0_SHARP),
				offValue: -2147483648,
				x: 100,
				y: 30,
				width: 10,
				height: 10,
			},
			{ id: 'note8', onValue: midiNoteToInt16(Note.G0), offValue: -2147483648, x: 110, y: 42, width: 10, height: 10 },
			{
				id: 'note9',
				onValue: midiNoteToInt16(Note.G0_SHARP),
				offValue: -2147483648,
				x: 120,
				y: 30,
				width: 10,
				height: 10,
			},
			{
				id: 'note10',
				onValue: midiNoteToInt16(Note.A0),
				offValue: -2147483648,
				x: 130,
				y: 42,
				width: 10,
				height: 10,
			},
			{
				id: 'note11',
				onValue: midiNoteToInt16(Note.A0_SHARP),
				offValue: -2147483648,
				x: 140,
				y: 30,
				width: 10,
				height: 10,
			},
			{
				id: 'note12',
				onValue: midiNoteToInt16(Note.B0),
				offValue: -2147483648,
				x: 150,
				y: 42,
				width: 10,
				height: 10,
			},
		],
		name: 'Quantizer',
		sliders: [],
		steppers: [],
	},
	cvToMidi: {
		width: 200,
		height: 100,
		connectors: [
			{ id: 'cvin', x: 5, y: 20, isInput: true, label: 'note in' },
			{ id: 'clockin', x: 5, y: 35, isInput: true, label: 'clock in' },
		],
		name: 'CV to MIDI',
		switches: [],
		sliders: [],
		steppers: [{ id: 'channel', x: 80, y: 10, width: 10, height: 20, minValue: 1, maxValue: 8 }],
		defaultValues: {
			channel: 1,
		},
	},
	saw: {
		width: 100,
		height: 100,
		connectors: [{ id: 'out', x: 85, y: 20, isInput: false }],
		name: 'Saw',
		switches: [],
		sliders: [{ id: 'rate', x: 10, y: 20, width: 10, height: 50, minValue: 0, maxValue: 2000, resolution: 10 }],
		defaultValues: {
			rate: 1000,
		},
		steppers: [],
	},
	triangle: {
		width: 100,
		height: 100,
		connectors: [{ id: 'out', x: 85, y: 20, isInput: false }],
		name: 'Triangle',
		switches: [],
		sliders: [],
		steppers: [],
	},
	randomGenerator: {
		width: 100,
		height: 100,
		connectors: [{ id: 'out', x: 85, y: 20, isInput: false }],
		name: 'Random',
		switches: [],
		sliders: [],
		steppers: [],
	},
	scope: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'in', x: 5, y: 45, isInput: true },
			{ id: 'out', x: 85, y: 45 },
		],
		name: 'Scope',
		switches: [],
		sliders: [],
		steppers: [],
	},
	attenuator: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'out', x: 85, y: 20, isInput: false },
			{ id: 'in', x: 5, y: 20, isInput: true },
		],
		name: 'Attenuator',
		switches: [],
		sliders: [{ id: 'divisor', x: 10, y: 20, width: 10, height: 50, minValue: 1, maxValue: 100, resolution: 1 }],
		defaultValues: {
			divisor: 1,
		},
		steppers: [],
	},
	offset: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'in', x: 5, y: 20, isInput: true },
			{ id: 'out', x: 85, y: 20, isInput: false },
		],
		name: 'Offset',
		sliders: [{ id: 'offset', x: 10, y: 20, width: 10, height: 50, minValue: 0, maxValue: 32000, resolution: 100 }],
		defaultValues: {
			offset: 0,
		},
		switches: [],
		steppers: [],
	},
	constant: {
		width: 100,
		height: 100,
		connectors: [{ id: 'out', x: 85, y: 20, isInput: false }],
		name: 'Constant',
		sliders: [{ id: 'out', x: 10, y: 20, width: 10, height: 50, minValue: -32000, maxValue: 32000, resolution: 100 }],
		defaultValues: {
			out: 0,
		},
		switches: [],
		steppers: [],
	},
	and: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'in1', x: 5, y: 20, isInput: true },
			{ id: 'in2', x: 5, y: 35, isInput: true },
			{ id: 'out', x: 85, y: 20, isInput: false },
		],
		name: 'And',
		sliders: [],
		switches: [],
		steppers: [],
	},
	or: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'in1', x: 5, y: 20, isInput: true },
			{ id: 'in2', x: 5, y: 35, isInput: true },
			{ id: 'out', x: 85, y: 20, isInput: false },
		],
		name: 'Or',
		sliders: [],
		switches: [],
		steppers: [],
	},
	xor: {
		width: 100,
		height: 100,
		connectors: [
			{ id: 'in1', x: 5, y: 20, isInput: true },
			{ id: 'in2', x: 5, y: 35, isInput: true },
			{ id: 'out', x: 85, y: 20, isInput: false },
		],
		name: 'Xor',
		sliders: [],
		switches: [],
		steppers: [],
	},
};
