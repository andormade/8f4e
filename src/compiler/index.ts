import {
	createFunctionSection,
	createTypeSection,
	createFunctionType,
	createExportSection,
	createFunctionExport,
	createCodeSection,
	createFunctionBody,
	createImportSection,
	createMemoryImport,
} from './wasm/sections';

import { call, i32store } from './wasm/instructions';
import { generateOutputAddressLookup } from './initializeMemory';
import * as moduleCompilers from './modules';
import { Module } from './modules/types';
import { i32abs, i32modulo } from './wasm/helpers';
import { Type } from 'wasm-bytecode-utils';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

const compileModules = function (modules): Module[] {
	let memoryAddress = 4;
	return modules
		.filter(({ type }) => moduleCompilers[type])
		.map(({ id, type }) => {
			const module = moduleCompilers[type](id, memoryAddress);
			memoryAddress += module.initialMemory.length * Int32Array.BYTES_PER_ELEMENT;
			return module;
		});
};

const generateMemoryInitiatorFunction = function (compiledModules) {
	return compiledModules
		.map(module => {
			let pointer = module.offset;
			return module.initialMemory
				.map(value => {
					const instuction = i32store(pointer, value);
					pointer += Int32Array.BYTES_PER_ELEMENT;
					return instuction;
				})
				.flat();
		})
		.flat();
};

const compile = function (modules: object[]) {
	const compiledModules = compileModules(modules);
	const functionBodies = compiledModules.map(({ functionBody }) => functionBody);
	const functionSignatures = compiledModules.map(() => 0x00);
	const cycleFunction = compiledModules.map((module, index) => call(index + 4)).flat();
	const memoryInitiatorFunction = generateMemoryInitiatorFunction(compiledModules);
	const outputAddressLookup = generateOutputAddressLookup(compiledModules);
	const helperFunctions = [i32abs(), i32modulo()];
	const helperFunctionSignatures = [0x01, 0x02];

	return {
		codeBuffer: Uint8Array.from([
			...HEADER,
			...VERSION,
			...createTypeSection([
				createFunctionType([], []),
				createFunctionType([Type.I32], [Type.I32]),
				createFunctionType([Type.I32, Type.I32], [Type.I32]),
			]),
			...createImportSection([createMemoryImport('js', 'memory', 1, 1, true)]),
			...createFunctionSection([...helperFunctionSignatures, 0x00, 0x00, ...functionSignatures]),
			...createExportSection([createFunctionExport('init', 0x02), createFunctionExport('cycle', 0x03)]),
			...createCodeSection([
				...helperFunctions,
				createFunctionBody([], memoryInitiatorFunction),
				createFunctionBody([], cycleFunction),
				...functionBodies,
			]),
		]),
		outputAddressLookup,
		compiledModules,
	};
};

export default compile;
