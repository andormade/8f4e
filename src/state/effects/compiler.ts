import { State } from '../types';
import { compilationDone, recompile } from '../mutators/compiler';

export default function compiler(state: State, events): void {
	const worker = new Worker(new URL('../../../packages/worker/src/index.ts', import.meta.url));
	// @ts-ignore shared: true
	const memoryRef = new WebAssembly.Memory({ initial: 1, maximum: 1, shared: true });

	async function onRecompile() {
		worker.postMessage({ memoryRef, modules: state.modules, connections: state.connections });
		recompile(state);
	}

	async function onWorkerMessage({ data }) {
		switch (data.type) {
			case 'midiMessage':
				events.dispatch('sendMidiMessage', data.payload);
				break;
			case 'compilationDone':
				compilationDone(state, data, memoryRef);
				state.modules.forEach(({ type }, index) => {
					if (state.moduleTypes[type].insertState) {
						state.moduleTypes[type].insertState(
							state.modules[index].state,
							state.compiler.memoryBuffer,
							state.compiler.memoryAddressLookup[state.modules[index].id]
						);
					}
				});
				break;
		}
	}

	worker.addEventListener('message', onWorkerMessage);
	events.on('createConnection', onRecompile);
	events.on('deleteConnection', onRecompile);
	events.on('addModule', onRecompile);
	events.on('deleteModule', onRecompile);
	events.on('init', onRecompile);
}
