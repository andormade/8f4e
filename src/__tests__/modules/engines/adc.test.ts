import { createTestModule } from '@8f4e/compiler';

import adc from '../../../modules/engines/adc.asm';
import { LOGIC_HIGH } from '../../../modules/engines/consts';

let testModule;

const fixtures = [
	[3000, LOGIC_HIGH],
	[2000, LOGIC_HIGH],
	[3000, LOGIC_HIGH],
];

describe('functional tests', () => {
	beforeAll(async () => {
		testModule = await createTestModule(adc(), { resolution: 8 });
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test.each(fixtures)('given %p the output should be %p', (input, output) => {
		const { memory, test } = testModule;

		memory[0] = input;
		test();
		expect(memory[2]).toBe(output);
	});
});
