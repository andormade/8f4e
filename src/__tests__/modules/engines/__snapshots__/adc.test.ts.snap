// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adc if the generated memory map matches with the snapshot 1`] = `
Map {
  "default" => Object {
    "byteAddress": 0,
    "default": 0,
    "relativeWordAddress": 0,
    "type": 5,
    "wordSize": 1,
  },
  "in" => Object {
    "byteAddress": 4,
    "default": 0,
    "relativeWordAddress": 1,
    "type": 3,
    "wordSize": 1,
  },
  "out:1" => Object {
    "byteAddress": 8,
    "default": 0,
    "relativeWordAddress": 2,
    "type": 4,
    "wordSize": 1,
  },
  "out:2" => Object {
    "byteAddress": 12,
    "default": 0,
    "relativeWordAddress": 3,
    "type": 4,
    "wordSize": 1,
  },
  "out:3" => Object {
    "byteAddress": 16,
    "default": 0,
    "relativeWordAddress": 4,
    "type": 4,
    "wordSize": 1,
  },
  "out:4" => Object {
    "byteAddress": 20,
    "default": 0,
    "relativeWordAddress": 5,
    "type": 4,
    "wordSize": 1,
  },
  "out:5" => Object {
    "byteAddress": 24,
    "default": 0,
    "relativeWordAddress": 6,
    "type": 4,
    "wordSize": 1,
  },
  "out:6" => Object {
    "byteAddress": 28,
    "default": 0,
    "relativeWordAddress": 7,
    "type": 4,
    "wordSize": 1,
  },
  "out:7" => Object {
    "byteAddress": 32,
    "default": 0,
    "relativeWordAddress": 8,
    "type": 4,
    "wordSize": 1,
  },
  "out:8" => Object {
    "byteAddress": 36,
    "default": 0,
    "relativeWordAddress": 9,
    "type": 4,
    "wordSize": 1,
  },
}
`;

exports[`adc if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32)
    i32.const 4
    i32.load
    i32.load
    i32.const 128
    i32.div_s
    local.set 0
    i32.const 8
    local.get 0
    i32.const 1
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 12
    local.get 0
    i32.const 2
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 16
    local.get 0
    i32.const 4
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 20
    local.get 0
    i32.const 8
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 24
    local.get 0
    i32.const 16
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 28
    local.get 0
    i32.const 32
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 32
    local.get 0
    i32.const 64
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store
    i32.const 36
    local.get 0
    i32.const 128
    i32.and
    i32.const 0
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 32767
    else
      i32.const 0
    end
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
